production/ksonnet/loki/memberlist.libsonnet

```diff
@@ -0,0 +1,141 @@
+  _config+:: {
+    gossip_member_label: 'loki_gossip_member',
+    // Enables use of memberlist for all rings, instead of consul. If multikv_migration_enabled is true, consul hostname is still configured,
+    // but "primary" KV depends on value of multikv_primary.
+    memberlist_ring_enabled: false,
+
+    // Migrating from consul to memberlist is a multi-step process:
+    // 1) Enable multikv_migration_enabled, with primary=consul, secondary=memberlist, and multikv_mirror_enabled=false, restart components.
+    // 2) Set multikv_mirror_enabled=true. This doesn't require restart.
+    // 3) Swap multikv_primary and multikv_secondary, ie. multikv_primary=memberlist, multikv_secondary=consul. This doesn't require restart.
+    // 4) Set multikv_migration_enabled=false and multikv_migration_teardown=true. This requires a restart, but components will now use only memberlist.
+    // 5) Set multikv_migration_teardown=false. This doesn't require a restart.
+    multikv_migration_enabled: false,
+    multikv_migration_teardown: false,
+    multikv_primary: 'consul',
+    multikv_secondary: 'memberlist',
+    multikv_switch_primary_secondary: false,
+    multikv_mirror_enabled: false,
+
+    multi_kv_obj: if $._config.multikv_migration_enabled then {
+      store: 'multi',
+      multi: {
+        primary: $._config.multikv_primary,
+        secondary: $._config.multikv_secondary,
+      },
+    } else {
+      store: 'memberlist',
+      consul: {
+        host: null,
+      },
+    },
+
+    loki+: {
+      ingester+: {
+        lifecycler+: {
+          ring+: {
+            kvstore+: $._config.multi_kv_obj,
+          },
+        },
+      },
+
+      distributor+: {
+        ring+: {
+          kvstore+: $._config.multi_kv_obj,
+        },
+      },
+
+      ruler+: if !$._config.ruler_enabled then {} else {
+        ring+: {
+          kvstore+: $._config.multi_kv_obj,
+        },
+      },
+
+      // other components pick this up from common
+      common+: {
+        ring+: {
+          kvstore+: $._config.multi_kv_obj,
+        },
+      },
+
+      memberlist: {
+        abort_if_cluster_join_fails: false,
+
+        // Expose this port on all distributor, ingester
+        // and querier replicas.
+        bind_port: gossipRingPort,
+
+        // You can use a headless k8s service for all distributor,
+        // ingester and querier components.
+        join_members: ['gossip-ring.%s.svc.cluster.local:%d' % [$._config.namespace, gossipRingPort]],
+
+        max_join_backoff: '1m',
+        max_join_retries: '10',
+        min_join_backoff: '1s',
+      },
+    },
+
+    // When doing migration via multi KV store, this section can be used
+    // to configure runtime parameters of multi KV store
+    multi_kv_config: if !$._config.multikv_migration_enabled && !$._config.multikv_migration_teardown then {} else {
+      primary: if $._config.multikv_switch_primary_secondary then $._config.multikv_secondary else $._config.multikv_primary,
```

Comment:
I think that we need to switch secondary at this time, otherwise, we will have `memberlist` as primary and as secondary :
```jsonnet
secondary: if $._config.multikv_switch_primary_secondary then $._config.multikv_primary, else $._config.multikv_secondary
```

Reply:
I can see how this would be confusing, it tripped me up at first as well. The actual config struct in the MultiKV for runtime reloading of the stores actually only allows for changing the Primary storage. But have a look at this function [here](https://github.com/grafana/dskit/blob/main/kv/multi.go#L154), if the primary changes then this code handles swapping the primary and secondary stores. So, we only need to set the primary and secondary initially, and then flip this jsonnet config that says we're doing a migration and want to swap the primary and secondaries. That will handle changing the primary to memberlist and the code internally will then swap the secondary to consul.